---
title: "Introduction to maseR"
author: "Diogo Veiga"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to maser}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

Alternative splicing affects 95% of human genes, and disregulation of this process is associated to several diseases. We developed the maseR package to facilitate the interpretation of splicing events generated by rMATS and to find isoforms affected by splicing events.

## Importing rMATS events
We demonstrate maseR using data from a study that measured hypoxia effects on a cancer cell line at 0h and 24h using RNA-seq (ref), and splicing events were detected using rMATS.  

We use `importEvents` indicating the folder path containing rMATS files, and labels for conditions. `importEvents` returns a list object that will be used throught the analysis containing genomic locations and PSI levels for all types of splicing events.

* SE, refers to exon skipping
* RI, intron retention
* MXE, mutually exclusive exons
* A3SS, alternative 3' splice site
* A5SS, alternative 5' splice site

```{r, warning = FALSE, message = FALSE}
library(maser)

# path to Hypoxia data
path <- system.file("extdata", file.path("MATS_output"),
                    package = "maser")
hypoxia <- maser(path, c("Hypoxia 0h", "Hypoxia 24h"))
hypoxia
```

We can remove low coverage events using `filterByCoverage`.
```{r, warning = FALSE, message = FALSE}
hypoxia_filt <- filterByCoverage(hypoxia, avg_reads = 5)
```

## Global overview of splicing changes
An overview of significant events can be obtained using either `dotplot` or `volcano` functions, specifying FDR levels, minimum change in PSI between conditions and the type of alternative splicing.

```{r, warning = FALSE, message = FALSE, fig.width = 7, fig.height = 5}
# dotplot(mats.filt, fdr = 0.05, deltaPSI = 0.1, type = "SE")
volcano(hypoxia_filt, fdr = 0.05, deltaPSI = 0.1, type = "SE")
```

The distribution of signficant events in each conditions can be plotted using `splicingDistribution` and desired significance thresholds. 
```{r, warning = FALSE, message = FALSE, fig.width = 7, fig.height = 3}
splicingDistribution(hypoxia_filt, fdr = 0.05, deltaPSI = 0.1)
```

Global splicing patterns of individual replicates can also be inspected using principal component analysis (PCA) or boxplots of PSI levels.

```{r, warning = FALSE, message = FALSE, fig.width = 6, fig.height = 5, fig.show='hold'}
pca(hypoxia_filt, type = "RI")
PSI_levels(hypoxia_filt, type = "RI")
```
The function `topEvents` allows to select statistically significant events given a FDR cutoff and minimum PSI change. Default values for FDR is 0.05 and deltaPSI is 0.1 (ie. 10% minimum change). 

```{r, warning = FALSE, message = FALSE, fig.width = 6, fig.height = 5, fig.show='hold'}
hypoxia_top <- topEvents(hypoxia_filt, fdr = 0.05, deltaPSI = 0.3 )
hypoxia_top
dotplot(hypoxia_top, fdr = 0.05, deltaPSI = 0.1, type = "SE")
```

Also, genes with frequent AS events can be discovered using `viewTopSplicedGenes`. The function rank genes based on the number of events, and a subset of AS types can be specificied. 
```{r, warning = FALSE, message = FALSE, fig.width = 3, fig.height = 4, fig.show='hold'}
viewTopSplicedGenes(hypoxia_top, n = 10)
viewTopSplicedGenes(hypoxia_top, n = 10, types = c("RI"))
```

## Gene-based analysis of splicing

We can select gene specific events using `geneEvents`. For instance, all AS events affecting MIB2 can be selected as below.
```{r, warning = FALSE, message = FALSE, fig.width = 3, fig.height = 4, fig.show='hold'}
hypoxia_mib2 <- geneEvents(hypoxia_filt, geneS = "MIB2", fdr = 0.05, deltaPSI = 0.1 )
print(hypoxia_mib2)
```

Next, PSI levels for exon skipping events affecting MIB2 can be plotted using `plotGenePSI`. 
```{r, warning = FALSE, message = FALSE, fig.width = 5, fig.height = 4, fig.show='hold'}
plotGenePSI(hypoxia_mib2, type = "SE", show_replicates = T)
```

**Code below needs description.**
```{r, warning = FALSE, message = FALSE, fig.width = 5, fig.height = 4, fig.show='hold'}

#Ensembl 85 - gene coordinates
ah = AnnotationHub::AnnotationHub()
ah = AnnotationHub::subset(ah,species=="Homo sapiens")
qhs <- AnnotationHub::query(ah, c("Ensembl", "gene", "annotation", "grch38"))
gtf <- qhs[["AH51014"]] #Homo_sapiens.GRCh38.85.gtf
gtf <- GenomeInfoDb::keepSeqlevels(gtf, c(paste0(seq(1:22)), "X", "Y"), pruning.mode = "coarse")
seqlevels(gtf, pruning.mode="coarse") <- c(paste0("chr", seq(1:22)), "chrX", "chrY")


#Importing GTF file
gtf_genc <- rtracklayer::readGFF("/Users/dveiga/tools/Gencode/gencode.v25.annotation.gtf")
gtf_genc_gr <- as(gtf_genc, "GRanges")

# analyzing exon skipping
hypoxia_gene <- geneEvents(hypoxia_filt, geneS = "TCF7L2", fdr = 0.05, deltaPSI = 0.1 )
plotTranscripts(hypoxia_gene, type = "SE", event_id = 37903, 
                gtf = gtf_genc_gr, is_strict = T, zoom = F)


#investigating intron retention
hypoxia_gene <- geneEvents(hypoxia_filt, geneS = "MIB2", fdr = 0.05, deltaPSI = 0.1 )
plotGenePSI(hypoxia_gene, type = "RI", show_replicates = T)
plotTranscripts(hypoxia_gene, type = "RI", event_id = 2508, 
                gtf = gtf_genc_gr, is_strict = T, zoom = T)

# investigating MXE
viewTopSplicedGenes(hypoxia_top, n = 10, types = c("MXE")) #should be replaced by display()
mats_gene <- geneEvents(hypoxia_filt, geneS = "TCF7L2", fdr = 0.05, deltaPSI = 0.1 )
plotGenePSI(mats_gene, type = "MXE", show_replicates = T)
plotTranscripts(mats_gene, type = "MXE", event_id = 2435, 
                gtf = gtf_genc_gr, is_strict = T, zoom = F)

# investigating A5SS
viewTopSplicedGenes(hypoxia.top, n = 10, types = c("A5SS"))
mats_gene <- geneEvents(hypoxia.filt, geneS = "BCS1L", fdr = 0.05, deltaPSI = 0.1 )
plotGenePSI(mats_gene, type = "A5SS", show_replicates = T)
plotTranscripts(mats_gene, type = "A5SS", event_id = 3988, 
                gtf = gtf_genc_gr, zoom = T)
#plotUniprotKBFeatures(mats_gene, "A5SS", 3988, gtf_genc_gr, features)

mats_gene <- geneEvents(mats.filt, geneS = "CEP170", fdr = 0.05, deltaPSI = 0.1 )
plotGenePSI(mats_gene, type = "A5SS", show_replicates = T)

plotTranscripts(mats_gene, type = "A5SS", event_id = 2511, 
                gtf = gtf_genc_gr, zoom = F)

# investigating A3SS
viewTopSplicedGenes(hypoxia_top, n = 10, types = c("A3SS"))
mats_gene <- geneEvents(hypoxia_filt, geneS = "STK36", fdr = 0.05, deltaPSI = 0.1 )
PSI_geneEvents(mats_gene, type = "A3SS", show_replicates = T)
plotTranscripts(mats_gene, type = "A3SS", event_id = 148, 
                gtf = gtf_genc_gr, zoom = T)

#negative strand test
mats_gene <- geneEvents(hypoxia_filt, geneS = "ELOVL1")
plotTranscripts(mats_gene, type = "A3SS", event_id = 3030, 
                gtf = gtf_genc_gr, zoom = T)

mats_gene <- geneEvents(hypoxia_filt, geneS = "RNF216")
plotTranscripts(mats_gene, type = "A3SS", event_id = 4224, 
                gtf = gtf_genc_gr, zoom = T)

mats_gene <- geneEvents(mats_filt, geneS = "SRSF5")
plotTranscripts(mats_gene, type = "A3SS", event_id = 2128, 
                gtf = gtf_genc_gr, zoom = T)

# Splicing events mapped to protein features

mats_gene <- geneEvents(hypoxia_filt, geneS = "MIB2", fdr = 0.05, deltaPSI = 0.1 )
availableTracksUniprotKB()
features <- c("domain", "transmem", "mutagen", "topo-dom", "motif")
plotUniprotKBFeatures(mats_gene, type = "RI", event_id = 2508, 
                      features = features, gtf = gtf_genc_gr, 
                      is_strict = T, zoom = F, 
                      show_transcripts = T)

mats_gene <- geneEvents(hypoxia_filt, geneS = "TCF7L2", fdr = 0.05, deltaPSI = 0.1 )
plotTranscripts(mats_gene, type = "SE", event_id = 37903, 
                gtf = gtf_genc_gr, is_strict = T, zoom = F)
plotUniprotKBFeatures(mats_gene, type = "SE", event_id = 37903, 
                      features = features, gtf = gtf_genc_gr, 
                      is_strict = T, zoom = F,
                      show_transcripts = T)

mats_gene <- geneEvents(hypoxia_filt, geneS = "STK36", fdr = 0.05, deltaPSI = 0.1 )
plotUniprotKBFeatures(mats_gene, type = "A3SS", event_id = 148, 
                      features = features, gtf = gtf_genc_gr, is_strict = T, zoom = F)

plotUniprotKBFeatures(mats_gene, "A3SS", 148, gtf_genc_gr, features)
```

