---
title: "Introduction"
shorttitle: "Introduction"
author: "Diogo Veiga"
package: maser
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc_float: true  
vignette: >
  %\VignetteIndexEntry{Introduction}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Overview of maser
Alternative splicing occurs in most of human genes and novel splice isoforms are associated to several diseases. Moreover, splicing events can affect both translation and function of proteins.  

We developed **maser** (**M**apping **A**lternative **S**plicing **E**vents to p**R**oteins), to integrate genomic information of splicing events to their corresponding protein features.  

In the basic workflow, `r Rpackage("maser")` imports events generated by rMATS and provides several types of plots and filtering functions for a through analysis of alternative splicing. The supported events are listed below.  


* SE, refers to exon skipping
* RI, intron retention
* MXE, mutually exclusive exons
* A3SS, alternative 3' splice site
* A5SS, alternative 5' splice site

The key feature of the package is mapping of splicing events to protein features such as topological domains, motifs and mutation sites provided by the [UniprotKB](http://www.uniprot.org/) database and visualized along side the genomic location of splicing.  

In this manner, `r Rpackage("maser")` can quickly identify splicing affecting known protein domains, extracellular and transmembrane regions, as well as mutation sites in the protein.  

# Importing rMATS events
We demonstrate the package with data generated to investigate the effects of hypoxia in alternative splicing in colorectal cancer <link>
[link](https://doi.org/10.1038/npjgenmed.2016.20). RNA-seq was collected at 0h and 24h after inducing hypoxia in the HCT116 colorectal cancer cell line, and splicing events were detected using rMATS.

To read in splicing events we use the constructor function `maser()` indicating the path containing rMATS files, and labels for conditions. The `maser()` constructor returns an object containing all events quantified by rMATS. 


```{r, warning = FALSE, message = FALSE}
library(maser)

# path to Hypoxia data
path <- system.file("extdata", file.path("MATS_output"),
                    package = "maser")
hypoxia <- maser(path, c("Hypoxia 0h", "Hypoxia 24h"))
```


```{r, warning = FALSE, message = FALSE}
hypoxia
```

Splicing events have genomic locations, raw counts, PSI levels and statistics (p-values) regarding their differential splicing between conditions. Access to different data types is done using `annot()`, `counts()`, `PSI()`, and `stats()`, which takes as argument the maser object and event type.

```{r, warning = FALSE, message = FALSE, eval=FALSE}
head(stats(hypoxia, type = "SE")[, 1:8])
```

```{r, warning = FALSE, message = FALSE, echo=FALSE}
knitr::kable(
  head(stats(hypoxia, type = "SE")[, 1:8])
)
```

# Filtering events
Low coverage splicing junctions are commonly found in RNA-seq data and lead to low confidence PSI levels. We can remove low coverage events using `filterByCoverage()`, which can signficantly reduced the number of splicing events.
```{r, warning = FALSE, message = FALSE}
hypoxia_filt <- filterByCoverage(hypoxia, avg_reads = 5)
```

The function `topEvents()` allows to select statistically significant events given a FDR cutoff and minimum PSI change. Default values are `fdr = 0.05` and `deltaPSI = 0.1` (ie. 10% minimum change). 

```{r, warning = FALSE, message = FALSE}
hypoxia_top <- topEvents(hypoxia_filt, fdr = 0.05, deltaPSI = 0.2)
hypoxia_top
```

Also, gene specific events can be selected using `geneEvents()`. For instance, there are 8 splicing changes affecting MIB2 as seen below.
```{r, warning = FALSE, message = FALSE}
hypoxia_mib2 <- geneEvents(hypoxia_filt, geneS = "MIB2", fdr = 0.05, deltaPSI = 0.1)
print(hypoxia_mib2)
```

PSI levels for gene events can be plotted using `plotGenePSI()`, indicating a valid splicing type.
```{r, warning = FALSE, message = FALSE}
plotGenePSI(hypoxia_mib2, type = "SE", show_replicates = T)
```

# Global splicing plots
An overview of significant events can be obtained using either `dotplot()` or `volcano()` functions, specifying FDR levels, minimum change in PSI between conditions and splicing type.
Significant events in each condition will be highlighted.

```{r, warning = FALSE, message = FALSE}
volcano(hypoxia_filt, fdr = 0.05, deltaPSI = 0.1, type = "SE")
```

If only significant events should be plotted, then use use `topEvents` to select events.
```{r, warning = FALSE, message = FALSE}
dotplot(hypoxia_top, type = "SE")
```

Splicing patterns of individual replicates can also be inspected using principal component analysis (PCA) or boxplots of PSI levels.

```{r, warning = FALSE, message = FALSE, fig.small = TRUE}
pca(hypoxia_filt, type = "RI")
PSI_levels(hypoxia_filt, type = "RI")
```

The breakdown of splicing types can be plotted using `splicingDistribution()` and desired significance thresholds. For instance, we can see the increase of intron retention after 24h of hypoxia.

```{r, warning = FALSE, message = FALSE}
splicingDistribution(hypoxia_filt, fdr = 0.05, deltaPSI = 0.1)
```


# Genomic visualization of splicing events

Users can use `r Rpackage("maser")` to visualize transcripts affected by splicing. The core function for genomic visualization is `plotTranscripts()`. This function will identify transcripts that match the pattern of the splicing variant using `r Biocpkg("GenomicRanges")`, and will use `r Biocpkg("Gviz")` for visualization of the location of splicing event.  

`plotTranscripts()` requires an Ensembl or Gencode GTF using the hg38 build of the human genome. Ensembl GTFs can be retrieved using `r Biocpkg("AnnotationHub")` as below. Several releases are available, and `r Rpackage("maser")` is compatible with any version using the hg38 build.  

Below is the snippet for retrieving Ensembl GTF release 85 using `r Biocpkg("AnnotationHub")`. 
```{r, warning = FALSE, message = FALSE}
ah <- AnnotationHub::AnnotationHub()
ah <- AnnotationHub::subset(ah, species == "Homo sapiens")
qhs <- AnnotationHub::query(ah, c("Ensembl", "gene", "annotation", "grch38"))
gtf <- qhs[["AH51014"]] #Homo_sapiens.GRCh38.85.gtf
```

Alternatively, GTFs can be downloaded from the [Ensembl FTP site](ftp://ftp.ensembl.org/pub/release-91/gtf/homo_sapiens) and imported using `readGFF()` from the `r Biocpkg("rtracklayer")` package.

Describe the use of `plotTranscripts()` for each splicing type, and zoom on and off.