---
title: "Mapping splice events to protein features"
shorttitle: "Mapping splice events to protein features"
author: "Diogo Veiga"
package: maser
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc_float: true  
vignette: >
  %\VignetteIndexEntry{Mapping splicing to protein features}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Introduction

The major functionality provided by `r Rpackage("maser")` is mapping of splicing events to [UniprotKB](http://www.uniprot.org/) protein features.  

Features such as protein domains, motifs and mutation sites can be annotated in batch for splicing events and visualized along side the genomic location of splicing.  


# Batch annotation of protein features

## Creating the maser object
We illustrate the workflow using the hypoxia dataset from the previous vignette.

```{r, warning = FALSE, message = FALSE}
library(maser)

# path to Hypoxia data
path <- system.file("extdata", file.path("MATS_output"),
                    package = "maser")
hypoxia <- maser(path, c("Hypoxia 0h", "Hypoxia 24h"))

# Remove low coverage events and get significant events
hypoxia_filt <- filterByCoverage(hypoxia, avg_reads = 5)
hypoxia_top <- topEvents(hypoxia_filt, deltaPSI = 0.2)

```

Available UniprotKB features can be queried using `availableFeaturesUniprotKB()`.

```{r, warning = FALSE, message = FALSE, eval=FALSE}
head(availableFeaturesUniprotKB(), 10)
```

```{r, warning = FALSE, message = FALSE, echo=FALSE}
knitr::kable(
  head(availableFeaturesUniprotKB(), 10)
)

```

## Mapping transcripts and protein to events 
The function `mapTranscriptsToEvents()` will add both transcript and protein IDs to all events in the maser object. First, transcripts overlapping exons involved in the the splice event are retrieved using `r Biocpkg("GenomicRanges")`. Then the transcript is mapped to its corresponding protein in Uniprot.  

`mapTranscriptsToEvents()` requires an Ensembl or Gencode GTF using the hg38 build of the human genome. Ensembl GTFs can be retrieved using `r Biocpkg("AnnotationHub")` package as below.

```{r, warning = FALSE, message = FALSE}
ah <- AnnotationHub::AnnotationHub()
ah <- AnnotationHub::subset(ah, species == "Homo sapiens")
qhs <- AnnotationHub::query(ah, c("Ensembl", "gene", "annotation", "grch38"))
ens_gtf <- qhs[["AH51014"]] #Homo_sapiens.GRCh38.85.gtf
```

```{r, warning = FALSE, message = FALSE}
hypoxia_top <- mapTranscriptsToEvents(hypoxia_top, gtf = ens_gtf)
```

If the mapping worked correctly, ENST and Uniprot identifiers will be added to events. `NA` values indicates non-protein coding transcripts.

```{r, warning = FALSE, message = FALSE, eval=FALSE}
head(annot(hypoxia_top, "SE"))
```

```{r, warning = FALSE, message = FALSE, echo=FALSE}
knitr::kable(
  head(annot(hypoxia_top, "SE"))
)
```

## Feature annotation

Now we are ready for batch annotation using `mapProteinFeaturesToEvents()`. The function retrieves the latest annotation from the public UniprotKB FTP, and overlaps splice events to protein features which have been mapped with hg38 genomic coordinates. UniprotKB features are grouped into categories.  

Multiple transcripts (and therefore multiple proteins) might be affected by the splice event. The annotation will indicate both the feature description and protein ID for which the annotation has been assigned. 

The batch annotation can be executed either by feature or category. If categories are provided, all features within the category group will be included for annotation. This may take a long time to process if dealing with a high number of splice events.

```{r, warning = FALSE, message = FALSE}
features <- c("domain", "transmem", "mutagen", "topo-dom", "motif", "signal", "DNA-bind")
categories_all <- c("Domain_and_Sites", "Mutagenesis", "Molecule_Processing",
                "PTM", "Structural_Features", "Topology")

categories <- c("Domain_and_Sites", "Mutagenesis", "Molecule_Processing", "Topology")

hypoxia_annot <- mapProteinFeaturesToEvents(hypoxia_top, categories, by = "category")

system.time({ hypoxia_annot <- mapProteinFeaturesToEvents(hypoxia_top, categories, by = "category") })

```

Feature annotation can be interactively displayed in a web browser using `display()` or retrieved as a `data.frame` using `annot()`.  

```{r, warning = FALSE, message = FALSE, eval = FALSE}
display(hypoxia_annot, "SE")
uniprot <- annot(hypoxia_annot, "SE")
```

```{r, warning = FALSE, message = FALSE}
table(uniprot$domain)
```

Cassette exon is in the kinase domain of RIPK2 (O43353). On the other hand, annotate mutations are in the upstream and downstream exons.


```{r, warning = FALSE, message = FALSE}

ripk2_events <- geneEvents(hypoxia_annot, "RIPK2")
plotUniprotKBFeatures(ripk2_events, type = "SE", event_id = 14319, 
                      features = c("domain", "mutagen"), gtf = ens_gtf, 
                      zoom = F,
                      show_transcripts = T)

plotUniprotKBFeatures(ripk2_events_annot, type = "SE", event_id = 14319, 
                      features = c("domain", "mutagen"), gtf = ens_gtf, 
                      zoom = F,
                      show_transcripts = T)

```

```{r, warning = FALSE, message = FALSE}

bmpr2_events <- geneEvents(hypoxia_annot, "BMPR2")
plotUniprotKBFeatures(bmpr2_events, type = "SE", event_id = 25501, 
                      features = c("domain", "topo-dom"), gtf = ens_gtf, 
                      zoom = F,
                      show_transcripts = T)

```






